### PR TITLE
inductor: add input type check for fuse_attention

### LINKED ISSUE #99124: [Inductor] [CPU] scaled_dot_product_attention() unexpected a value type caused crash  in xcit_large_24_p8_224
### 🐛 Describe the bug

Found in https://github.com/pytorch/pytorch/issues/93531#issuecomment-1507847237 **(Single-core Single-thread)**

Error message:

```bash
cpu  eval  xcit_large_24_p8_224                [2023-04-13 18:22:45,243] torch._inductor.utils: [WARNING] make_fallback(aten.cumprod): a decomposition exists, we should switch to it
ERROR:common:Backend dynamo failed in warmup()
Traceback (most recent call last):
  File "/workspace/pytorch/benchmarks/dynamo/common.py", line 1365, in warmup
    fn(model, example_inputs)
  File "/workspace/pytorch/torch/_dynamo/eval_frame.py", line 254, in _fn
    return fn(*args, **kwargs)
  File "benchmarks/dynamo/timm_models.py", line 327, in forward_pass
    return mod(*inputs)
  File "/workspace/pytorch/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/workspace/pytorch/torch/_dynamo/eval_frame.py", line 401, in catch_errors
    return callback(frame, cache_size, hooks)
  File "/workspace/pytorch/torch/_dynamo/convert_frame.py", line 474, in _convert_frame
    result = inner_convert(frame, cache_size, hooks)
  File "/workspace/pytorch/torch/_dynamo/convert_frame.py", line 118, in _fn
    return fn(*args, **kwargs)
  File "/workspace/pytorch/torch/_dynamo/convert_frame.py", line 317, in _convert_frame_assert
    return _compile(
  File "/workspace/pytorch/torch/_dynamo/utils.py", line 169, in time_wrapper
    r = func(*args, **kwargs)
  File "/workspace/pytorch/torch/_dynamo/convert_frame.py", line 382, in _compile
    out_code = transform_code_object(code, transform)
  File "/workspace/pytorch/torch/_dynamo/bytecode_transformation.py", line 683, in transform_code_object
    transformations(instructions, code_options)
  File "/workspace/pytorch/torch/_dynamo/convert_frame.py", line 369, in transform
    tracer.run()
  File "/workspace/pytorch/torch/_dynamo/symbolic_convert.py", line 1892, in run
    super().run()
  File "/workspace/pytorch/torch/_dynamo/symbolic_convert.py", line 611, in run
    and self.step()
  File "/workspace/pytorch/torch/_dynamo/symbolic_convert.py", line 571, in step
    getattr(self, inst.opname)(inst)
  File "/workspace/pytorch/torch/_dynamo/symbolic_convert.py", line 1979, in RETURN_VALUE
    self.output.compile_subgraph(
  File "/workspace/pytorch/torch/_dynamo/output_graph.py", line 622, in compile_subgraph
    self.compile_and_call_fx_graph(tx, list(reversed(stack_values)), root)
  File "/workspace/pytorch/torch/_dynamo/output_graph.py", line 692, in compile_and_call_fx_graph
    compiled_fn = self.call_user_compiler(gm)
  File "/workspace/pytorch/torch/_dynamo/utils.py", line 169, in time_wrapper
    r = func(*args, **kwargs)
  File "/workspace/pytorch/torch/_dynamo/output_graph.py", line 774, in call_user_compiler
    raise BackendCompilerFailed(self.compiler_fn, e).with_traceback(
  File "/workspace/pytorch/torch/_dynamo/output_graph.py", line 770, in call_user_compiler
    compiled_fn = compiler_fn(gm, self.fake_example_inputs())
  File "/workspace/pytorch/torch/_dynamo/debug_utils.py", line 1098, in debug_wrapper
    compiled_gm = compiler_fn(gm, example_inputs)
  File "/workspace/pytorch/torch/_dynamo/backends/inductor.py", line 9, in inductor
    return compile_fx(*args, **kwargs)
  File "/workspace/pytorch/torch/_inductor/compile_fx.py", line 736, in compile_fx
    return aot_autograd(
  File "/workspace/pytorch/torch/_dynamo/backends/common.py", line 62, in compiler_fn
    cg = aot_module_simplified(gm, example_inputs, **kwargs)
  File "/workspace/pytorch/torch/_functorch/aot_autograd.py", line 3093, in aot_module_simplified
    compiled_fn = create_aot_dispatcher_function(
  File "/workspace/pytorch/torch/_dynamo/utils.py", line 169, in time_wrapper
    r = func(*args, **kwargs)
  File "/workspace/pytorch/torch/_functorch/aot_autograd.py", line 2728, in create_aot_dispatcher_function
    compiled_fn = compiler_fn(flat_fn, fake_flat_args, aot_config, fw_metadata=fw_metadata)
  File "/workspace/pytorch/torch/_functorch/aot_autograd.py", line 1827, in aot_wrapper_dedupe
    return compiler_fn(flat_fn, leaf_flat_args, aot_config, fw_metadata=fw_metadata)
  File "/workspace/pytorch/torch/_functorch/aot_autograd.py", line 1993, in aot_wrapper_synthetic_base
    return compiler_fn(flat_fn, flat_args, aot_config, fw_metadata=fw_metadata)
  File "/workspace/pytorch/torch/_functorch/aot_autograd.py", line 1292, in aot_dispatch_base
    compiled_fw = compiler(fw_module, flat_args)
  File "/workspace/pytorch/torch/_dynamo/utils.py", line 169, in time_wrapper
    r = func(*args, **kwargs)
  File "/workspace/pytorch/torch/_inductor/compile_fx.py", line 685, in fw_compiler_base
    joint_graph_passes(model)
  File "/workspace/pytorch/torch/_inductor/fx_passes/joint_graph.py", line 28, in joint_graph_passes
    if patterns.apply(graph.graph):
  File "/workspace/pytorch/torch/_inductor/pattern_matcher.py", line 591, in apply
    if m and entry.extra_check(m):
  File "/workspace/pytorch/torch/_inductor/pattern_matcher.py", line 509, in check_fn
    match.replacement_graph = trace_fn(replace_fn, args)
  File "/workspace/pytorch/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/workspace/pytorch/torch/_inductor/pattern_matcher.py", line 673, in inference_graph
    gm = make_fx(fn, select_decomp_table())(*args)
  File "/workspace/pytorch/torch/fx/experimental/proxy_tensor.py", line 761, in wrapped
    t = dispatch_trace(wrap_key(func, args, fx_tracer, pre_autograd), tracer=fx_tracer, concrete_args=tuple(phs))
  File "/workspace/pytorch/torch/_dynamo/eval_frame.py", line 254, in _fn
    return fn(*args, **kwargs)
  File "/workspace/pytorch/torch/fx/experimental/proxy_tensor.py", line 467, in dispatch_trace
    graph = tracer.trace(root, concrete_args)
  File "/workspace/pytorch/torch/_dynamo/eval_frame.py", line 254, in _fn
    return fn(*args, **kwargs)
  File "/workspace/pytorch/torch/fx/_symbolic_trace.py", line 778, in trace
    (self.create_arg(fn(*args)),),
  File "/workspace/pytorch/torch/fx/experimental/proxy_tensor.py", line 484, in wrapped
    out = f(*tensors)
  File "/workspace/pytorch/torch/_inductor/fx_passes/fuse_attention.py", line 47, in _sfdp_replacement_2
    return aten.scaled_dot_product_attention(
  File "/workspace/pytorch/torch/_ops.py", line 646, in __call__
    return self._op(*args, **kwargs or {})
  File "/workspace/pytorch/torch/_inductor/overrides.py", line 33, in __torch_function__
    return replace_fn(func)(*args, **kwargs)
  File "/workspace/pytorch/torch/_ops.py", line 646, in __call__
    return self._op(*args, **kwargs or {})
torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
RuntimeError: aten::scaled_dot_product_attention() Expected a value of type 'Optional[float]' for argument 'scale' but instead found type 'FakeTensor'.
Position: 6
Value: FakeTensor(FakeTensor(..., device='meta', size=(16, 1, 1)), cpu)
Declaration: aten::scaled_dot_product_attention(Tensor query, Tensor key, Tensor value, Tensor? attn_mask=None, float dropout_p=0., bool is_causal=False, *, float? scale=None) -> Tensor
Cast error details: Unable to cast Python instance to C++ type (#define PYBIND11_DETAILED_ERROR_MESSAGES or compile in debug mode for details)


You can suppress this exception and fall back to eager by setting:
    torch._dynamo.config.suppress_errors = True

ERROR


```

SW information:

SW	| Nightly commit	| Master/Main commit
-- | -- | --
Pytorch|[84ec5d9](https://github.com/pytorch/pytorch/commit/84ec5d9)|[6ff32b5](https://github.com/pytorch/pytorch/commit/6ff32b5)
Torchbench|/|[83a316df](https://github.com/pytorch/benchmark/commit/83a316df)
torchaudio|[375e751](https://github.com/pytorch/audio/commit/375e751)|[a8f4e97](https://github.com/pytorch/audio/commit/a8f4e97)
torchtext|[9749082](https://github.com/pytorch/text/commit/9749082)| [46e7eef](https://github.com/pytorch/text/commit/46e7eef)
torchvision|[8d15ca7](https://github.com/pytorch/vision/commit/8d15ca7)|[98c5815](https://github.com/pytorch/vision/commit/98c5815)
torchdata|[b3048d5](https://github.com/pytorch/data/commit/b3048d5)|[f1283eb](https://github.com/pytorch/data/commit/f1283eb)
dynamo_benchmarks|[1238ae3](https://github.com/pytorch/pytorch/commit/1238ae3)|/

log with `TORCH_COMPILE_DEBUG=1` :
[single_thread_xcit_large_24_p8_224_rerun__bench_20230414_023546.log](https://github.com/pytorch/pytorch/files/11228384/single_thread_xcit_large_24_p8_224_rerun__bench_20230414_023546.log)

Repro:

```bash
python -m torch.backends.xeon.run_cpu --core_list 0 --ncores_per_instance 1 benchmarks/dynamo/timm_models.py--performance --float32 -dcpu -n50 --inductor  --no-skip --dashboard --only xcit_large_24_p8_224   --cold_start_latency --batch_size 1 --threads 1

```

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### PR DESCRIPTION
**Pull request by XiaobingSuper**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #99502
* __->__ #99296

For TIMM ```xcit_large_24_p8_224```, the scale factor is a tensor(https://github.com/huggingface/pytorch-image-models/blob/main/timm/models/xcit.py#L205), and ```scaled_dot_product_attention``` doesn't support it, this PR will add a check which only does the fusion when the scale factor is float/int value.

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: test/inductor/test_fused_attention.py [hunk 1]
```diff
@@ -165,6 +165,42 @@ def sfdp_pattern_6(query, key, value):
 
         self._check_common(sfdp_pattern_6, contains=False)
 
+    def test_pattern_fails_with_tensor_factor(self):
+        # https://github.com/pytorch/pytorch/issues/99124
```
Comment (XiaobingSuper) [inline test/inductor/test_fused_attention.py:169]: This issue is reported on CPU side, but I checked it is also happening on GPU side, so add this case here.

File: torch/_inductor/fx_passes/fuse_attention.py [hunk 2]
```diff
@@ -141,6 +146,44 @@ def _sfdp_replacement_6(query, key, value, attn_mask, dropout_p):
...
+        view_node = list(matmuls[0].users.keys())[0]
+        if (
+            len(view_node.users) != 1
+            or list(view_node.users.keys())[0].target != scale_factor_op
+        ):
+            return False
+        scale_factor_node = list(view_node.users.keys())[0]
+        if len(scale_factor_node.args) != 2:
+            return False
+        # make sure the scale_factor a float/int. SymInt?
```
Comment (EikanWang) [inline torch/_inductor/fx_passes/fuse_attention.py:170]: Should we add a dynamic shape case to verify `SymInt`?

File: torch/_inductor/fx_passes/fuse_attention.py [hunk 3]
```diff
@@ -141,6 +146,44 @@ def _sfdp_replacement_6(query, key, value, attn_mask, dropout_p):
 # TODO(jansel): make these pattern work with lowmem_dropout=True
 
 
+def _sfdp_scale_factor_check(scale_factor_op):
+    def fn(match):
+        # make sure the mul(div) for the scale factor is scalar mul(div).
+        # bmm->view->mul(div)
+        matmuls = filter_nodes(match.nodes, aten.bmm)
+        if len(matmuls) < 2:
```
Comment (jansel) [inline torch/_inductor/fx_passes/fuse_attention.py:154]: Shouldn't this already be checked by the pattern?  Can you give an example and add a test for when this fails?

File: torch/_inductor/fx_passes/fuse_attention.py [hunk 4]
```diff
@@ -141,6 +146,44 @@ def _sfdp_replacement_6(query, key, value, attn_mask, dropout_p):
...
+            return False
+        if (
+            len(matmuls[0].users) != 1
```
Comment (jansel) [inline torch/_inductor/fx_passes/fuse_attention.py:157]: Shouldn't this already be checked by the pattern?  Can you give an example and add a test for when this fails?

File: torch/_inductor/fx_passes/fuse_attention.py [hunk 5]
```diff
@@ -141,6 +146,44 @@ def _sfdp_replacement_6(query, key, value, attn_mask, dropout_p):
...
+            or list(matmuls[0].users.keys())[0].target != aten.view.default
+        ):
+            return False
+        view_node = list(matmuls[0].users.keys())[0]
+        if (
+            len(view_node.users) != 1
```
Comment (jansel) [inline torch/_inductor/fx_passes/fuse_attention.py:163]: Shouldn't this already be checked by the pattern?  Can you give an example and add a test for when this fails?

File: torch/_inductor/fx_passes/fuse_attention.py [hunk 6]
```diff
@@ -141,6 +146,44 @@ def _sfdp_replacement_6(query, key, value, attn_mask, dropout_p):
...
+            or list(view_node.users.keys())[0].target != scale_factor_op
+        ):
+            return False
+        scale_factor_node = list(view_node.users.keys())[0]
```
Comment (jansel) [inline torch/_inductor/fx_passes/fuse_attention.py:167]: Could you go directly here with `filter_nodes(match.nodes, scale_factor_op)`?

File: test/inductor/test_fused_attention.py [hunk 7]
```diff
@@ -165,6 +165,42 @@ def sfdp_pattern_6(query, key, value):
...
+                    y = y.mul(self.scale_factor)
+                return y.softmax(dim=-1).matmul(value)
+
+        tensor_shape = (2, 4, 4, 4)
+        args = [
+            torch.randn(tensor_shape, device="cuda"),
+            torch.randn(tensor_shape, device="cuda"),
+            torch.randn(tensor_shape, device="cuda"),
+        ]
+        with torch.no_grad():
```
Comment (jansel) [inline test/inductor/test_fused_attention.py:192]: Can you test training as well?

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_inductor/fx_passes/fuse_attention.py (hunk 3 above)
Comment (XiaobingSuper) [reply torch/_inductor/fx_passes/fuse_attention.py:154]: Yes, this has been checked by the pattern, I do that to make sure it is all right. I will remove it to simplify it.

File: torch/_inductor/fx_passes/fuse_attention.py (hunk 2 above)
Comment (XiaobingSuper) [reply torch/_inductor/fx_passes/fuse_attention.py:170]: Yes, we may need to add dynamic shape case, but the patterns doesn't match for dynamic shape case. Let do it at next step.

File: torch/_inductor/fx_passes/fuse_attention.py (hunk 4 above)
Comment (XiaobingSuper) [reply torch/_inductor/fx_passes/fuse_attention.py]: removed such a check.

File: torch/_inductor/fx_passes/fuse_attention.py (hunk 5 above)
Comment (XiaobingSuper) [reply torch/_inductor/fx_passes/fuse_attention.py]: removed such a check.

File: torch/_inductor/fx_passes/fuse_attention.py (hunk 6 above)
Comment (XiaobingSuper) [reply torch/_inductor/fx_passes/fuse_attention.py]: Yes, changed.

File: test/inductor/test_fused_attention.py (hunk 7 above)
Comment (XiaobingSuper) [reply test/inductor/test_fused_attention.py]: training test is added, please note that even if not have this PR, the training path can work well(the pattern doesn't match), and has an accuracy gap compared with eager mode.

**Review by jgong5 — APPROVED** [review]

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**